### ISSUE TITLE
setjmp and EMULATE_FUNCTION_POINTER_CASTS=1

### ISSUE THREAD

**Issue by mflatt** [emscripten-core/emscripten#16177]
**Version of emscripten/emsdk:**
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.2-git
clang version 14.0.0 (https://github.com/llvm/llvm-project.git 782c0dd1a1c235afb09a34e7da4a1267ead14765)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /opt/homebrew/opt/emscripten/libexec/llvm/bin

**Failing steps:**
Using this "demo.c":
```
#include <stdio.h>
#include <setjmp.h>

int main(int argc, const char *argv[]) {
  jmp_buf jb;

  switch (setjmp(jb)) {
  case 0:
    printf("ok\n");
    break;
  default:
    printf("done\n");
    break;
  }
  return 0;
}
```

Compilation and linking succeed:

```
laptop$ emcc -s EMULATE_FUNCTION_POINTER_CASTS=1 demo.c 
```

Running fails:

```
laptop$ node a.out.js
/private/tmp/a.out.js:147
      throw ex;
      ^

TypeError: Cannot convert 1078 to a BigInt
    at invoke_iii (/private/tmp/a.out.js:2004:34)
    at wasm://wasm/00017ade:wasm-function[10]:0x450
    at /private/tmp/a.out.js:1557:22
    at callMain (/private/tmp/a.out.js:2293:15)
    at doRun (/private/tmp/a.out.js:2350:23)
    at run (/private/tmp/a.out.js:2365:5)
    at runCaller (/private/tmp/a.out.js:2271:19)
    at removeRunDependency (/private/tmp/a.out.js:1466:7)
    at receiveInstance (/private/tmp/a.out.js:1646:5)
    at receiveInstantiationResult (/private/tmp/a.out.js:1663:5)
```

When I compile my actual application in certain modes, such as when it uses lots of `uint64_t`s instead of `uint32_t`s, then `setjmp` and `EMULATE_FUNCTION_POINTER_CASTS=1` seem to get along ok.

**Comment by sbc100** [comment]
Thats odd.. I wonder why BigInt is being used at all here seeing as you are not building with `WASM_BIGINT`.

**Comment by kripken** [comment]
That error message is odd. Maybe it depends on which node? I get `TypeError: wasm function signature contains illegal type`...

Fix + more details in #16183
